src/crimson/admin/pg_commands.cc

```diff
@@ -0,0 +1,159 @@
+// -*- mode:C++; tab-width:8; c-basic-offset:2; indent-tabs-mode:t -*-
+// vim: ts=8 sw=2 smarttab
+
+#include "crimson/admin/pg_commands.h"
+
+#include <memory>
+#include <string>
+#include <string_view>
+
+#include <fmt/format.h>
+#include <seastar/core/future.hh>
+
+#include "crimson/admin/admin_socket.h"
+#include "crimson/osd/osd.h"
+#include "crimson/osd/pg.h"
+
+
+using crimson::osd::OSD;
+using crimson::osd::PG;
+using namespace crimson::common;
+
+
+namespace crimson::admin::pg {
+
+class PGCommand : public AdminSocketHook {
+public:
+  // TODO: const correctness of osd
+  PGCommand(crimson::osd::OSD& osd,
+            std::string_view prefix,
+            std::string_view desc,
+            std::string_view help)
+      : AdminSocketHook{prefix, desc, help}, osd {osd}
+  {}
+  seastar::future<tell_result_t> call(const cmdmap_t& cmdmap,
+                                      std::string_view format,
+                                      ceph::bufferlist&& input) const final
+  {
+    // we have "ceph tell <pgid> <cmd>". and it is the ceph cli's responsibility
+    // to add "pgid" to the cmd dict. as rados_pg_command() does not set it for
+    // us. moreover, and "pgid" is not listed in the command description, as user
+    // command format does not follow the convention of "<prefix> [<args>,...]"
+    // so we have to verify it on the server side.
+    std::string pgid_str;
+    pg_t pgid;
+    if (!cmd_getval(cmdmap, "pgid", pgid_str)) {
+      return seastar::make_ready_future<tell_result_t>(
+        tell_result_t{-EINVAL, "no pgid specified"});
+    } else if (!pgid.parse(pgid_str.c_str())) {
+      return seastar::make_ready_future<tell_result_t>(
+        tell_result_t{-EINVAL, fmt::format("couldn't parse pgid '{}'", pgid_str)});
+    }
+    // am i the primary for this pg?
+    const auto osdmap = osd.get_shard_services().get_osdmap();
+    spg_t spg_id;
+    if (!osdmap->get_primary_shard(pgid, &spg_id)) {
+      return seastar::make_ready_future<tell_result_t>(tell_result_t{
+          -ENOENT, fmt::format("pgid '{}' does not exist", pgid_str)});
+    }
+    Ref<PG> pg = osd.get_pg(spg_id);
+    if (!pg) {
+      return seastar::make_ready_future<tell_result_t>(tell_result_t{
+        -ENOENT, fmt::format("i don't have pgid '{}'", spg_id)});
+    }
+    if (!pg->is_primary()) {
+      return seastar::make_ready_future<tell_result_t>(tell_result_t{
+        -EAGAIN, fmt::format("not primary for pgid '{}'", spg_id)});
+    }
+    return this->do_command(pg, cmdmap, format, std::move(input));
+  }
+
+private:
+  virtual seastar::future<tell_result_t>
+  do_command(Ref<PG> pg,
+             const cmdmap_t& cmdmap,
+             std::string_view format,
+             ceph::bufferlist&& input) const = 0;
+
+  OSD& osd;
+};
+
+class QueryCommand final : public PGCommand {
```

Comment:
`QueryCommand` is not a `PGCommand`, but it is a `AdminSocketHook`. i could use CRTP to make this fact more obvious. but that'd be overkill.

Reply:
Adding a comment?

Reply:
i'd prefer not adding a comment. i think it's more confusing if i failed to send the right message.